File: erpc_python/erpc/crc16.py
#!/usr/bin/env python

# Copyright (c) 2016 Freescale Semiconductor, Inc.
# Copyright 2016-2025 NXP
#
# SPDX-License-Identifier: BSD-3-Clause

from typing import Union

class Crc16:
    POLY = 0x1021

    def __init__(self, crcStart: int=0xEF4A):
        self._crc_start = crcStart
        self._table = [self.compute_table(i) for i in range(256)]

    def compute_table(self, i: int):
        crc = 0
        i <<= 8
        for _ in range(8):
            temp = crc ^ i
            crc <<= 1
            if temp & 0x8000:
                crc ^= self.POLY
            i <<= 1
        return crc

    def compute_crc16(self, data: Union[str, bytes]) -> int:
        if isinstance(data, str):
            data = bytes(data.encode())
        crc = self._crc_start
        for c in data:
            crc = ((crc << 8) ^ self._table[((crc >> 8) ^ c) & 0xff]) & 0xffff
        return crc

File: erpc_java/src/main/java/io/github/embeddedrpc/erpc/auxiliary/Crc16.java
/*
 * Copyright 2023 NXP
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

package  io.github.embeddedrpc.erpc.auxiliary;

/**
 * Class for computing ITU-CCITT CRC-16.
 */
public class Crc16 {
    private static final int POLY = 0x1021;
    private final int crcStart;
    private final int[] table;

    /**
     * CRC16 constructor.
     *
     * @param crcStart CRC16 start position
     */
    @SuppressWarnings("magicnumber")
    public Crc16(int crcStart) {
        this.crcStart = crcStart;
        this.table = new int[256];
        for (int i = 0; i < 256; i++) {
            this.table[i] = this.computeTable(i);
        }
    }

    /**
     * CRC16 constructor with start position 0xEF4A.
     */
    @SuppressWarnings("magicnumber")
    public Crc16() {
        this(0xEF4A);
    }

    /**
     * Function return CRC16 for given byte array.
     *
     * @param data Input byte array
     * @return Computed CRC16
     */
    @SuppressWarnings("magicnumber")
    public int computeCRC16(byte[] data) {
        int crc = this.crcStart;
        for (byte c : data) {
            crc = ((crc << 8) ^ this.table[((crc >> 8) ^ c) & 0xFF]) & 0xFFFF;
        }
        return crc;
    }

    @SuppressWarnings("magicnumber")
    private int computeTable(int i) {
        int crc = 0;
        int y = i;
        y <<= 8;
        for (int b = 0; b < 8; b++) {
            int temp = crc ^ y;
            crc <<= 1;
            if ((temp & 0x8000) != 0) {
                crc ^= POLY;
            }
            y <<= 1;
        }
        return crc;
    }
}
